component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -0,0 +1,222 @@
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.oauth.common.OAuth2ErrorCodes;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.bean.OAuthClientAuthnContext;
+import org.wso2.carbon.identity.oauth2.client.authentication.AbstractOAuthClientAuthenticator;
+import org.wso2.carbon.identity.oauth2.client.authentication.OAuthClientAuthnException;
+import org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.validator.JWTValidator;
+
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.http.HttpServletRequest;
+
+import static org.apache.commons.lang.StringUtils.isEmpty;
+import static org.apache.commons.lang.StringUtils.isNotEmpty;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.DEFAULT_VALIDITY_PERIOD_IN_MINUTES;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.OAUTH_JWT_ASSERTION;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.OAUTH_JWT_ASSERTION_TYPE;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.OAUTH_JWT_BEARER_GRANT_TYPE;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.PREVENT_TOKEN_REUSE;
+
+/**
+ * Client Authentication handler to implement oidc private_key_jwt client authentication spec
+ * http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication.
+ */
+public class PrivateKeyJWTClientAuthenticator extends AbstractOAuthClientAuthenticator {
+
+    private static final Log log = LogFactory.getLog(PrivateKeyJWTClientAuthenticator.class);
+
+    JWTValidator jwtValidator;
+
+    public PrivateKeyJWTClientAuthenticator() {
+
+        jwtValidator = createJWTValidator();
+    }
+
+    /**
+     * To check whether the authentication is successful.
+     *
+     * @param httpServletRequest      http servelet request
+     * @param content                 map of request body params
+     * @param oAuthClientAuthnContext oAuthClientAuthnContext
+     * @return true if the authentication is successful.
+     * @throws OAuthClientAuthnException
+     */
+    @Override
+    public boolean authenticateClient(HttpServletRequest httpServletRequest, Map<String, List> content,
+                                      OAuthClientAuthnContext oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        String clientAssertion = null;
+        List oauthJWTAssertion = content.get(OAUTH_JWT_ASSERTION);
+        if (oauthJWTAssertion != null && oauthJWTAssertion.get(0) != null) {
+            clientAssertion = oauthJWTAssertion.get(0).toString();
+        }
+        try {
+            boolean isValidAssertion = jwtValidator.isValidAssertion(getSignedJWT(clientAssertion));
```

Comment:
I think we should not solely depend on proxy logic. Ie if someone else comes and call this method it should properly authenticate client. But in this case if the client id is not in context it will return false. That is because you have a presumption that client id was set from somewhere else

Reply:
If it's set from jwtValidator.isValidAssertion method, it's theoretically not meaningful. ie while validating we can't expect it it to set client id to context